### PR TITLE
BUG: fix yaml table serialization compatibility with numpy 2

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Combines the regression test from #16414 and cherry-picks https://github.com/astropy/astropy/pull/15065/commits/9cdd95d51911d099bddedf0747f8eb40d78f219a and 3be8af07e23d9daab97533a991495cb9d315bfbc from #15065

This doesn't break any test that I could run locally against numpy 1.26, so it should be backportable, but I'm still keeping it as a draft until I see CI go green, just in case.

Fix https://github.com/astropy/astropy/issues/15792

Also resolves ~90% failures (340/380) from tests that currently rely on legacy print options as setup in
https://github.com/astropy/astropy/blob/b2e38be321ac6fb0c844704cbd4be1d1bf511b9b/astropy/conftest.py#L28

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/io/misc/tests/test_yaml.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py:321]: oh my, how did I miss this... thank you so much !

**Comment by pllim** [comment]
Diff LGTM but I'll let subpackage maintainer(s) review and approve. Thanks!

File: astropy/io/misc/yaml.py
Comment (astrofrog) [inline astropy/io/misc/yaml.py:249]: Just to make sure I understand, why not np.isnan?

Comment (neutrinoceros) [reply astropy/io/misc/yaml.py:249]: I assume that's for performance reasons, though, that's a question for @mhvk 
(`math.isnan` should also be faster than `np.isnan` for scalars)

File: astropy/io/misc/tests/test_yaml.py
Comment (mhvk) [inline astropy/io/misc/tests/test_yaml.py:322]: I'm confused why our regular tests didn't catch this. We test a whole slew of different `SkyCoord` for all serializations (see `astropy/io/tests/mixin_columns.py`) - was the problem specific to `frame="galactic"`? If so, perhaps add an extra `SkyCoord` in `mixin_columns`? (Though note that one then needs to add quite a bit of detail further down in the file...) Though in a way also fine just as is - after all, it is the incantation that proved there was a problem.

File: astropy/io/misc/yaml.py
Comment (mhvk) [reply astropy/io/misc/yaml.py:249]: Hmm, I vaguely recall that I just copied the snipped from how pyyaml does it, except changing `repr(data)` to `str(data)` (I *clearly* should have added a note with that...). I'm not quite sure why they worked this way - perhaps there are python float-like's that do not support `isnan`?

**Review by mhvk — COMMENTED** [review]
Beyond the two in-line comments, shouldn't we also get the next commit, for `complex`? 6b7916d4e3736eaac73ca25372c341079ae38291

File: astropy/io/misc/yaml.py
Comment (astrofrog) [reply astropy/io/misc/yaml.py:249]: Ok no problem - if we can get things to work with ``math.isnan`` I think that would be preferable

Comment (neutrinoceros) [reply astropy/io/misc/yaml.py:249]: For the record here's the line you apparently copied https://github.com/yaml/pyyaml/blob/48838a3c768e3d1bcab44197d800145cfd0719d6/lib/yaml/representer.py#L172
It was written 15 years ago (circa 2009), while `math.isnan` was introduced in Python 2.6, released in October 2008, so it's possible that the only reason they didn't use it was backward compatibility, and it just never got refactored.
Since the whole function is copied from pyyaml, it seems best to keep it as is, but I'll add a note so this knowledge doesn't get lost.

File: astropy/io/misc/tests/test_yaml.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py:322]: The problem is actually *not* specific to `frame="galactic"`, so I'll drop this argument to avoid confusion.
According to its docstring, `astropy/io/tests/mixin_columns.py` is used in `ascii/tests/test_ecsv.py`, `fits/tests/test_connect.py`, and `misc/tests/test_hdf5.py`, but not in yaml tests (yet). I'll see if there are any problems with using it here too.

Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py]: I've wrote a new test based on `test_hdf5.py::test_hdf5_mixins_qtable_to_table`, which also fails on main and pass with this branch. I ditched the first test, let me know if I should keep both.

File: astropy/io/misc/yaml.py
Comment (neutrinoceros) [reply astropy/io/misc/yaml.py:251]: FYI @mhvk I also updated your comment. It said that numpy scalars failed on "not NUMPY_LT_1_20" which I assume was a mistake and you really meant "not NUMPY_LT_2_0"

**Comment by neutrinoceros** [comment]
>  shouldn't we also get the next commit, for complex? 

It doesn't hurt and is clearly correct, but I should point out that it's also not tested. I'll push it now.

File: astropy/io/misc/tests/test_hdf5.py
Comment (taldcroft) [inline astropy/io/misc/tests/test_hdf5.py:710]: Why this change?

File: astropy/io/misc/tests/test_yaml.py
Comment (taldcroft) [inline astropy/io/misc/tests/test_yaml.py:322]: This regression test is fine but more importantly there should be YAML-level unit tests covering all the cases in the new code. I.e. serializing a simple dict structure with all the impacted data types and confirming the output and correct round-trip.

File: astropy/io/misc/tests/test_hdf5.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_hdf5.py:710]: wow, that's a syntaxically valid mistake I must have made by trying to open vscode from another editor and typing in the wrong tab. Thanks for spotting it

File: astropy/io/misc/tests/test_yaml.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py:345]: Putting it here is consistent with how the "inspiration" test lives in `test_hdf5.py`. Moving that one would be out of scope here but maybe we can move both in a follow up PR ?

File: astropy/io/misc/tests/test_yaml.py
Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py:322]: Gotcha. I'll work on it.

Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py:331]: Actually the unit tests in question already exists, I just needed to add a pytest fixture to compensate for global settings (see #15096).
They also capture the regressions properly, so maybe we should just ditch the more complex integration test, if that's okay with you.

Comment (taldcroft) [reply astropy/io/misc/tests/test_yaml.py:331]: Excellent about the tests already being there!  And yes, I'm good with dropping the integration test.

Comment (neutrinoceros) [reply astropy/io/misc/tests/test_yaml.py]: great, I rebased and greatly simplified the branch !

File: astropy/io/misc/yaml.py
Comment (mhvk) [reply astropy/io/misc/yaml.py:251]: Thanks, @neutrinoceros, that all makes sense.

**Review by mhvk — APPROVED** [review]
Ah, thanks, now I understand why the tests didn't have a problem before - it was the use of legacy printing. With this, it looks all great!

**Comment by mhvk** [comment]
Sorry, @taldcroft, I merged perhaps too quickly! A problem with trying to do a few things while really not at work; should just have approved and left it. Though hopefully you were OK with the latest push too...

**Comment by taldcroft** [comment]
@mhvk @neutrinoceros - Sometimes codecov seems to give false reports, but here I don't see any coverage of the `inf`, `nan`, or `1e17` cases. Those would be good, so a follow-up PR would be a good thing.

**Comment by neutrinoceros** [comment]
added to my todo list for tomorrow !

**Comment by pllim** [comment]
Thanks, all! I opened a follow-up issue at https://github.com/astropy/astropy/issues/16429

**Pull request merged**
